### PR TITLE
gh-94182: run the PidfdChildWatcher on the running loop

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-94182

### CODE REVIEW

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]: this test wasn't enabled?

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion
        asyncio.set_child_watcher(asyncio.PidfdChildWatcher())
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion
                *PROGRAM_CAT,
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion
            watcher = mock.create_autospec(asyncio.AbstractChildWatcher)
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion
            asyncio.set_child_watcher(watcher)
            self.assertIs(asyncio.get_child_watcher(), watcher)
            self.assertFalse(asyncio.get_child_watcher().is_active())
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]: this was patching the wrong attribute - asyncio calls `asyncio.get_child_watcher().__enter__().is_active()`

**Comment by graingert** [comment]
@asvetlov can I get a review on this please?

File: Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst]:
```suggestion
run the :class:`asyncio.PidfdChildWatcher` on the running loop
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: I split this fix out, so this PR is easier to review: https://github.com/python/cpython/pull/95009

File: Lib/test/test_asyncio/test_subprocess.py
Comment (agronholm) [inline Lib/test/test_asyncio/test_subprocess.py]: Maybe separate different logical blocks with blank lines.

File: Lib/test/test_asyncio/test_subprocess.py
Comment (agronholm) [inline Lib/test/test_asyncio/test_subprocess.py]: `asyncio.new_event_loop` is the default value, so why pass it at all?

File: Lib/test/test_asyncio/test_subprocess.py
Comment (agronholm) [inline Lib/test/test_asyncio/test_subprocess.py]: While this technically works, I would much rather see this made into a module level function.

File: Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst
Comment (agronholm) [inline Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst]: Maybe elaborate a little bit more? This is not very informative.

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: Ah well about that...

**Review by agronholm — CHANGES_REQUESTED** [review]
Some of my previous comments request changes.

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: This is extremely confusing asyncio behavior that I strongly disagree with https://github.com/python/cpython/pull/94593/files#r914528655

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: I'm copying the existing code here

https://github.com/python/cpython/pull/94184/files#diff-3a12ea0e8988b2edcb96962049a9cca2db1ee02db133cd99493ccc87a81d08e2R686

Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: When I move it to a module level flag I'll apply this fix

File: Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst
Comment (graingert) [inline Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst]:
```suggestion
run the :class:`asyncio.PidfdChildWatcher` on the running loop, this allows event loops to run subprocesses when there is no default event loop running on the main thread
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [reply Lib/test/test_asyncio/test_subprocess.py]: passing the loop factory disables the calls to set the `asyncio.set_event_loop()`

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion
@functools.cache
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (graingert) [inline Lib/test/test_asyncio/test_subprocess.py]:
```suggestion

if __name__ == '__main__':
```

File: Lib/test/test_asyncio/test_subprocess.py
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_subprocess.py:35]: This same function is also introduced by @kumaraditya303's PR GH-98024 (which uses PidfdChildWatcher when supported). I propose to refactor things so that everything uses Kumar's version. We can do that in Kumar's PR once this PR is merged.

File: Lib/asyncio/unix_events.py
Comment (gvanrossum) [inline Lib/asyncio/unix_events.py:935]: Curious why we can't pass the loop as an argument as well? (Though maybe that creates an unhealthy self-reference to the loop.)

**Review by gvanrossum — APPROVED** [review]
LGTM. We'll probably merge this ASAP.

**Comment by gvanrossum** [comment]
(Waiting for @1st1 review and passing tests.)

**Review by 1st1 — APPROVED** [review]

**Comment by gvanrossum** [comment]
Not sure what's wrong with the docs build.

File: Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst
Comment (1st1) [inline Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst:1]:
```suggestion
run the ``asyncio.PidfdChildWatcher`` on the running loop, this allows event loops to run subprocesses when there is no default event loop running on the main thread
```

**Comment by 1st1** [comment]
> Not sure what's wrong with the docs build.

I've fixed it.

File: Lib/asyncio/unix_events.py
Comment (graingert) [reply Lib/asyncio/unix_events.py:935]: Can probably skip storing the loop on `self` and use `get_running_loop` every time

File: Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst
Comment (1st1) [inline Misc/NEWS.d/next/Library/2022-06-24-08-49-47.gh-issue-94182.Wknau0.rst]:
```suggestion
run the :class:`asyncio.PidfdChildWatcher` on the running loop, this allows event loops to run subprocesses when there is no default event loop running on the main thread
```

**Comment by 1st1** [comment]
> I've fixed it.

Reverted the "fix" because apparently it was caused by an unrelated commit in main.

**Comment by JelleZijlstra** [comment]
I can take care of getting CI green and hitting the merge button here.

**Comment by 1st1** [comment]
> I can take care of getting CI green and hitting the merge button here.

Please do. You might need to rebase this PR to the latest main.

**Comment by gvanrossum** [comment]
The macOS build hit some GitHub Actions limit. I'm going to merge without waiting for it.

**Pull request merged**
